### DISCUSSION TITLE
Problem to reconnecting and play music to bleutooth speaker

### DISCUSSION THREAD

**Discussion by nadmaga** [pschatzmann/ESP32-A2DP#439] [Q&A] [answered]
Hi,
Firstly, I would like to thank you for all the amazing work!!
I try to send music from i2s to a bleutooth speaker (JBL Clip3) with an ESP32 (i2s source is another ESP32). I used the "base-player-a2dp.ino" example and replaced the player with the StreamCopy. Everything works fine and I can hear the music on the speaker.

My issue is, that sometimes it takes long until I hear the music after paaring (3-4 seconds). This is for me ok, but when I turn off the speaker and turn it on again, then sometims nothing happen, sometimes it takes 15 to 20 seconds until the music is streamed. Does this related to the buffer or is this an bluethooth issue of the ESP32?

I try also to use the 'set_ssid_callback' and disabled the 'set_auto_reconnect(false)' to connect to a bluetooth device. But the callback function is only called once at the beginning after reboot the ESP32. When I turn off and on the speaker again, it reconnect it, but without the call of the callback function 'isValid'. Do I somthing wrong? How can I improove the connection and the speed? Thank you very much for any inputs.

Here my code:

```
/**
 * @file base-player-a2dp.ino
 * @author Phil Schatzmann
 * @brief Sketch which uses the A2DP callback to provide data from the AudioPlayer via a Queue
 * 
 * @version 0.1
 * @date 2022-12-04
 * 
 * @copyright Copyright (c) 2022
 * 
 */

#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"

int buffer_count = 30;
int buffer_size = 512;

I2SStream i2sStream;                            // Access I2S as stream
BluetoothA2DPSource a2dp;

SynchronizedBufferRTOS<uint8_t> buffer(buffer_size*buffer_count, buffer_size, portMAX_DELAY, 10); // fast synchronized buffer
QueueStream<uint8_t> out(buffer); // convert Buffer to Stream
StreamCopy copier(out, i2sStream); // copy i2sStream to a2dpStream

// Provide data to A2DP
int32_t get_data(uint8_t *data, int32_t bytes) {
   size_t result_bytes = buffer.readArray(data, bytes);
   return result_bytes;
}

void connection_state_changed(esp_a2d_connection_state_t state, void *ptr){
  Serial.println(a2dp.to_str(state));
  if(state == ESP_A2D_CONNECTION_STATE_DISCONNECTED){
    Serial.println("==> DISCONNECTED");
  }
  if(state == ESP_A2D_CONNECTION_STATE_CONNECTED){
    Serial.println("==> CONNECTED");
  }
}

// Return true to connect, false will continue scanning
bool isValid(const char* ssid, esp_bd_addr_t address, int rssi){
   Serial.print("available SSID: ");
   Serial.println(ssid);
   if( ssid == "JBL Clip 3" or "Sennheiser MB Pro 1"){
     return true;
   }else{
     return false;
   }
}

void setup() {
 Serial.begin(115200);
 AudioLogger::instance().begin(Serial, AudioLogger::Warning);

 // start QueueStream
 out.begin();

 // start i2s input with default configuration
 Serial.println("starting I2S...");
 auto cfg = i2sStream.defaultConfig(RX_MODE);
  cfg.i2s_format = I2S_STD_FORMAT; // or try with I2S_LSB_FORMAT
  cfg.bits_per_sample = 16;
  cfg.channels = 2;
  cfg.sample_rate = 44100;
  cfg.is_master = false;
 i2sStream.begin(cfg);

 // start a2dp source
 Serial.println("starting A2DP...");
 a2dp.set_auto_reconnect(false);
 a2dp.set_on_connection_state_changed(connection_state_changed);
 a2dp.set_ssid_callback(isValid);
 a2dp.start_raw(get_data);  
// a2dp.set_volume(30);

 Serial.println("Started!");
}

void loop() {
 // decode data to buffer
 if(a2dp.is_connected()){
   copier.copy();
 }
}

```

**Comment by pschatzmann** [comment, marked as answer]
It is my understanding that the issue is with your bluetooth speaker which does not accept the new connection and does not recognize that the old connection was lost.

This issue should go away if you end the bluetooth properly before rebooting the ESP32.

You can set the Core Debug Level to Info or Debug to reproduce what's going on in detail

**Reply by nadmaga** [reply]
Wow thank you for the fast reply.
No, the speaker connects to the esp after I turn the speaker on again (connection_state_changed is called with the new state ESP_A2D_CONNECTION_STATE_CONNECTED). but the callback function 'isValid' is not called and no sound is playing (the getData callback function is not called).
With end up the bluetooth properly, do you mean to call a2dp.end()? When should I do it in my code?

**Comment by pschatzmann** [comment]
That's my point: the speaker does not allow any connections until you restart it again! That confirms that the issue is with the speaker.
Since there is nothing like a shutdown hook on the ESP32 you need to create your own shutdown or restart button that would call end()

**Comment by lurcullu01** [comment]
Hey @nadmaga! I wanted to DM you but GitHub doesn't have that feature. I'm actually working on a very similar project but I'm facing some issues sending the data from ESP to ESP. To my understanding, you were able to get this done, Do you mind sharing your code? I also see you're using RTOS, which could also serve me useful in the future. Thanks!

**Comment by pschatzmann** [comment]
From the examples directory you should be able to take any receiver sketch and sender sketch. Just make sure that the names are matching! And be careful with the automatic reconnect. I suggest to switch it off for both!

To analyse what is happening in detail, just activate the logging...
